core/vm/contracts.go

```diff
@@ -360,140 +361,245 @@ type bigModExp struct {
+	u256_berlinDivisor    = uint256.NewInt(berlinDivisor)
+	u256_osakaDivisor     = uint256.NewInt(osakaDivisor)
 )
 
-// modexpMultComplexity implements bigModexp multComplexity formula, as defined in EIP-198
+// byzantiumMultComplexity implements bigModexp multComplexity formula, as defined in EIP-198
 //
 //	def mult_complexity(x):
 //		if x <= 64: return x ** 2
 //		elif x <= 1024: return x ** 2 // 4 + 96 * x - 3072
 //		else: return x ** 2 // 16 + 480 * x - 199680
 //
 // where is x is max(length_of_MODULUS, length_of_BASE)
-func modexpMultComplexity(x *big.Int) *big.Int {
+func byzantiumMultComplexity(x uint64) *uint256.Int {
 	switch {
-	case x.Cmp(big64) <= 0:
-		x.Mul(x, x) // x ** 2
-	case x.Cmp(big1024) <= 0:
-		// (x ** 2 // 4 ) + ( 96 * x - 3072)
-		x = new(big.Int).Add(
-			new(big.Int).Rsh(new(big.Int).Mul(x, x), 2),
-			new(big.Int).Sub(new(big.Int).Mul(big96, x), big3072),
-		)
+	case x <= 64:
+		return new(uint256.Int).SetUint64(x * x)
+	case x <= 1024:
+		// x^2 / 4 + 96*x - 3072
+		result := x*x/4 + 96*x - 3072
+		return new(uint256.Int).SetUint64(result)
 	default:
-		// (x ** 2 // 16) + (480 * x - 199680)
-		x = new(big.Int).Add(
-			new(big.Int).Rsh(new(big.Int).Mul(x, x), 4),
-			new(big.Int).Sub(new(big.Int).Mul(big480, x), big199680),
-		)
+		// For large x, use uint256 arithmetic to avoid overflow
+		// x^2 / 16 + 480*x - 199680
+		xUint := new(uint256.Int).SetUint64(x)
+
+		// Calculate x^2
+		xSquared := new(uint256.Int).Mul(xUint, xUint)
+
+		// Calculate x^2 / 16 (right shift by 4 bits)
+		xSquaredDiv16 := new(uint256.Int).Rsh(xSquared, 4)
+
+		// Calculate 480 * x
+		x480 := new(uint256.Int).Mul(xUint, u256_480)
+
+		// Calculate 480 * x - 199680
+		x480Minus199680 := new(uint256.Int).Sub(x480, u256_199680)
+
+		// Add the two parts together
+		return new(uint256.Int).Add(xSquaredDiv16, x480Minus199680)
+	}
+}
+
+// berlinMultComplexity implements the multiplication complexity formula for Berlin.
+//
+// def mult_complexity(x):
+//
+//	ceiling(x/8)^2
+//
+// where is x is max(length_of_MODULUS, length_of_BASE)
+func berlinMultComplexity(x uint64) *uint256.Int {
+	// TODO: The preceding line is too smart.
+	// TODO: The issue is that (x+7) / 8 can overflow
+	// TODO: if x > 2^64 - 7
+	ceilDiv8 := (x >> 3) + ((x&7 + 7) >> 3) // safe ceil(x / 8)
+	z := new(uint256.Int).SetUint64(ceilDiv8)
+	return new(uint256.Int).Mul(z, z) // square without overflow
+}
+
+// Slow Bigint way (benchmark this)
+// func berlinMultComplexity(xInt uint64) *big.Int {
+// 	x := new(big.Int).SetUint64(xInt)
+// 	x = new(big.Int).Add(x, big.NewInt(7))       // x + 7
+// 	x = new(big.Int).Rsh(x, 3)          // (x + 7) / 8
+// 	return new(big.Int).Mul(x, x)       // ((x + 7) / 8) ^ 2
+// }
+
+// osakaMultComplexity implements the multiplication complexity formula for Osaka.
+//
+// For x <= 32: returns 16
+// For x > 32: returns 2 * ceiling(x/8)^2
+func osakaMultComplexity(x uint64) *uint256.Int {
+	if x <= 32 {
+		return u256_16
```

Comment:
Check this: might want to allocate a new integer here since we are returning a pointer